File: sim/flysim.py
"""Bit-exact Python replica of FlyBrain.sol's simulation.

Used to calibrate the dynamics parameters before deployment and to generate
differential-test fixtures for the Solidity tests. Every operation mirrors the
contract: int32 arithmetic with truncating division, keccak256(step) noise,
synchronous spikes, engram plasticity at the end of each tick, walking.
"""
import json, os, struct, math
from Crypto.Hash import keccak

ROOT = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..')
COS16 = [125, 106, 71, 25, -25, -71, -106, -125, -125, -106, -71, -25, 25, 71, 106, 125]
SIN16 = [25, 71, 106, 125, 125, 106, 71, 25, -25, -71, -106, -125, -125, -106, -71, -25]
CH_NONE, CH_CUE, CH_TURN_LEFT, CH_TURN_RIGHT, CH_SHOCK = 0, 1, 2, 3, 4
T_EPG, T_EPGT, T_PEG, T_PEN_A, T_PEN_B, T_DELTA7 = range(6)
BIAS_MAX = 24
STRIDE = 16
WEDGES = 16

DEFAULT_PARAMS = dict(leak=120, thresh=1000, reset=-300, vMin=-2000, gains=[40, 40, 40, 40, 40, -40], gBias=8,
                      noise=120, stimGain=40, stimTTL=64, walkThreshold=100, maxSteps=64)


def keccak256(b: bytes) -> int:
    k = keccak.new(digest_bits=256); k.update(b); return int.from_bytes(k.digest(), 'big')


def tdiv(a, b):  # Solidity int division truncates toward zero
    q = abs(a) // abs(b)
    return q if (a >= 0) == (b > 0) else -q


def isqrt(x):
    if x == 0: return 0
    z = (x + 1) // 2; y = x
    while z < y:
        y = z; z = (x // z + z) // 2
    return y


class Circuit:
    def __init__(self, table: bytes):
        assert table[0] == 1
        self.N = table[1]; self.S = struct.unpack('>H', table[2:4])[0]
        N, S = self.N, self.S
        o = 4
        self.type = list(table[o:o + N]); o += N
        self.wedge = list(table[o:o + N]); o += N
        self.side = list(table[o:o + N]); o += N
        self.offsets = [struct.unpack('>H', table[o + 2 * i:o + 2 * i + 2])[0] for i in range(N + 1)]; o += 2 * (N + 1)
        self.syn = [(table[o + 2 * j], table[o + 2 * j + 1]) for j in range(S)]; o += 2 * S
        self.root = [struct.unpack('>Q', table[o + 8 * i:o + 8 * i + 8])[0] for i in range(N)]
        self.out = [self.syn[self.offsets[i]:self.offsets[i + 1]] for i in range(N)]

    @classmethod
    def load(cls, path=None):
        path = path or os.path.join(ROOT, 'contracts', 'data', 'circuit.hex')
        return cls(bytes.fromhex(open(path).read().strip()[2:]))


class FlyBrain:
    def __init__(self, circuit: Circuit, params=None, energy=10**9):
        self.c = circuit; self.p = dict(DEFAULT_PARAMS, **(params or {}))
        N = circuit.N
        self.v = [0] * N; self.bias = [0] * N; self.inp = [0] * N
        self.hist = [0] * WEDGES
        self.step = 0; self.energy = energy; self.alive = True; self.generation = 0
        self.posX = 0; self.posY = 0; self.headX = 0; self.headY = 0
        self.stimChannel = CH_NONE; self.stimParam = 0; self.stimStrength = 0; self.stimUntil = 0
        self.totalSpikes = 0
        self.log = []  # per-step (spikes, epg spikes by wedge)

    # ------------------------------------------------------------ actions
    def stimulate(self, channel, param, strength):
        assert self.alive and 1 <= channel <= 4 and 1 <= strength <= 255
        self.stimChannel, self.stimParam, self.stimStrength = channel, param, strength
        self.stimUntil = self.step + self.p['stimTTL']

    def _build_stim(self):
        c, p = self.c, self.p
        amp = self.stimStrength * p['stimGain']
        stim = [0] * c.N
        ch, param = self.stimChannel, self.stimParam
        for i in range(c.N):
            t = c.type[i]
            if ch == CH_CUE:
                if t <= T_EPGT and c.wedge[i] != 255:
                    dist = (c.wedge[i] + WEDGES - param) % WEDGES
                    if dist == 0: stim[i] = amp
                    elif dist in (1, WEDGES - 1): stim[i] = tdiv(amp, 2)
            elif ch in (CH_TURN_LEFT, CH_TURN_RIGHT):
                if t in (T_PEN_A, T_PEN_B) and ((ch == CH_TURN_LEFT and c.side[i] == 0) or (ch == CH_TURN_RIGHT and c.side[i] == 1)):
                    stim[i] = amp
            elif ch == CH_SHOCK:
                if t == T_DELTA7: stim[i] = amp
        return stim

    def tick(self, steps):
        assert self.alive and 1 <= steps <= self.p['maxSteps']
        c, p = self.c, self.p
        N = c.N
        stimActive = self.stimChannel != CH_NONE and self.step < self.stimUntil
        stimI = self._build_stim() if stimActive else [0] * N
        spk = [0] * N; bins = [0] * WEDGES; hx = hy = 0; tickSpikes = 0
        s0 = self.step; ran = 0
        while ran < steps and self.energy > 0:
            sn = s0 + ran
            stimNow = stimActive and sn < self.stimUntil
            rnd = keccak256(sn.to_bytes(8, 'big'))
            spikeList = []
            for i in range(N):
                if (i & 31) == 0 and i != 0:
                    rnd = keccak256(rnd.to_bytes(32, 'big'))
                nz = ((rnd >> ((i & 31) * 8)) & 0xFF) - 128
                x = self.v[i]
                x -= tdiv(x * p['leak'], 1024)
                x += self.inp[i] + tdiv(nz * p['noise'], 128) + self.bias[i] * p['gBias']
                if stimNow: x += stimI[i]
                self.inp[i] = 0
                if x < p['vMin']: x = p['vMin']
                if x >= p['thresh']:
                    x = p['reset']
                    spikeList.append(i); spk[i] += 1
                    if c.type[i] <= T_EPGT and c.wedge[i] != 255:
                        hx += COS16[c.wedge[i]]; hy += SIN16[c.wedge[i]]; bins[c.wedge[i]] += 1
                self.v[i] = x
            tickSpikes += len(spikeList)
            stepbins = [0] * WEDGES
            for i in spikeList:
                g = p['gains'][c.type[i]]
                for post, w in c.out[i]:
                    self.inp[post] += tdiv(w * g, 16)
                if c.type[i] <= T_EPGT and c.wedge[i] != 255: stepbins[c.wedge[i]] += 1
            self.log.append((len(spikeList), stepbins))
            ran += 1; self.energy -= 1
        # plasticity
        for i in range(N):
            b = self.bias[i]
            if spk[i] * 8 >= ran:
                if b < BIAS_MAX: b += 1
            elif spk[i] == 0:
                if b > -BIAS_MAX: b -= 1
            self.bias[i] = b
        # heading memory + walk
        best, bestCount = 0, 0
        for w in range(WEDGES):
            if bins[w] > bestCount: bestCount, best = bins[w], w
        if bestCount > 0 and self.hist[best] < 0xFFFF: self.hist[best] += 1
        mag = isqrt(hx * hx + hy * hy)
        if mag > 0 and mag >= p['walkThreshold'] * ran:
            self.posX += tdiv(hx * STRIDE * ran, mag)
            self.posY += tdiv(hy * STRIDE * ran, mag)
        self.headX, self.headY = hx, hy
        self.last_spk = spk
        self.step = s0 + ran
        self.totalSpikes += tickSpikes
        if stimActive and self.step >= self.stimUntil: self.stimChannel = CH_NONE
        if self.energy == 0: self.alive = False
        return dict(steps=ran, spikes=tickSpikes, headX=hx, headY=hy, posX=self.posX, posY=self.posY)

    # ------------------------------------------------------------ analysis
    def bump(self, last=1):
        """Population vector of EPG activity over the last `last` logged steps: (angle_rad, magnitude, total_epg_spikes)."""
        bins = [0] * WEDGES
        for _, sb in self.log[-last:]:
            for w in range(WEDGES): bins[w] += sb[w]
        x = sum(bins[w] * math.cos((w + 0.5) * 2 * math.pi / WEDGES) for w in range(WEDGES))
        y = sum(bins[w] * math.sin((w + 0.5) * 2 * math.pi / WEDGES) for w in range(WEDGES))
        tot = sum(bins)
        return math.atan2(y, x), math.hypot(x, y), tot

    def state_dict(self):
        return dict(v=list(self.v), bias=list(self.bias), hist=list(self.hist), step=self.step, energy=self.energy,
                    posX=self.posX, posY=self.posY, headX=self.headX, headY=self.headY, alive=self.alive)


if __name__ == '__main__':
    c = Circuit.load()
    print('N', c.N, 'S', c.S, 'types', {t: c.type.count(t) for t in range(6)})
    b = FlyBrain(c)
    b.stimulate(CH_CUE, 4, 4)
    for k in range(8):
        r = b.tick(32)
        a, m, tot = b.bump(8)
        print(f'tick {k}: spikes={r["spikes"]:4d} bump angle={math.degrees(a):6.1f}° mag={m:6.1f} epg={tot:4d} pos=({r["posX"]},{r["posY"]})')

File: web/src/lib/flysim.ts
// @ts-nocheck
// Bit-exact TypeScript port of FlyBrain.sol's simulation (see sim/flysim.py).
import { ethers } from "ethers";

export const COS16 = [125, 106, 71, 25, -25, -71, -106, -125, -125, -106, -71, -25, 25, 71, 106, 125];
export const SIN16 = [25, 71, 106, 125, 125, 106, 71, 25, -25, -71, -106, -125, -125, -106, -71, -25];
export const WEDGES = 16, BIAS_MAX = 24, STRIDE = 16;
export const CH = { NONE: 0, CUE: 1, TURN_LEFT: 2, TURN_RIGHT: 3, SHOCK: 4 };
export const T = { EPG: 0, EPGT: 1, PEG: 2, PEN_A: 3, PEN_B: 4, DELTA7: 5 };
export const TYPE_NAMES = ["EPG", "EPGt", "PEG", "PEN_a", "PEN_b", "Δ7"];
export const CH_NAMES = ["none", "cue", "turn left", "turn right", "shock"];

const tdiv = (a: number, b: number) => Math.trunc(a / b);
function isqrt(x: number) { if (x === 0) return 0; let z = Math.trunc((x + 1) / 2), y = x; while (z < y) { y = z; z = Math.trunc((Math.trunc(x / z) + z) / 2); } return y; }
const keccakU64 = (n: number) => BigInt(ethers.keccak256(ethers.zeroPadValue(ethers.toBeHex(n), 8)));
const keccakU256 = (b: bigint) => BigInt(ethers.keccak256(ethers.zeroPadValue(ethers.toBeHex(b), 32)));

export interface Params { leak: number; thresh: number; reset: number; vMin: number; gains: number[]; gBias: number; noise: number; stimGain: number; stimTTL: number; walkThreshold: number; maxSteps: number; persistInput?: boolean; }

export class Circuit {
  N: number; S: number; type: number[]; wedge: number[]; side: number[]; out: [number, number][][]; root: string[];
  constructor(hex: string) {
    const d = ethers.getBytes(hex);
    if (d[0] !== 1) throw new Error("circuit table version");
    const N = d[1], S = (d[2] << 8) | d[3];
    let o = 4;
    this.N = N; this.S = S;
    this.type = Array.from(d.slice(o, o + N)); o += N;
    this.wedge = Array.from(d.slice(o, o + N)); o += N;
    this.side = Array.from(d.slice(o, o + N)); o += N;
    const off: number[] = []; for (let i = 0; i <= N; i++) off.push((d[o + 2 * i] << 8) | d[o + 2 * i + 1]); o += 2 * (N + 1);
    this.out = [];
    for (let i = 0; i < N; i++) { const lst: [number, number][] = []; for (let j = off[i]; j < off[i + 1]; j++) lst.push([d[o + 2 * j], d[o + 2 * j + 1]]); this.out.push(lst); }
    o += 2 * S;
    this.root = []; for (let i = 0; i < N; i++) { let v = 0n; for (let k = 0; k < 8; k++) v = (v << 8n) | BigInt(d[o + 8 * i + k]); this.root.push(v.toString()); }
  }
}

export interface ChainState { v: number[]; bias: number[]; hist: number[]; pendingInput?: number[]; step: number; energy: number; alive: boolean; generation: number; posX: number; posY: number; headX: number; headY: number; stim?: { channel: number; param: number; strength: number; untilStep: number }; }

export class FlySim {
  c: Circuit; p: Params; v: number[]; bias: number[]; inp: number[]; hist: number[];
  step = 0; energy = 0; alive = true; generation = 0; posX = 0; posY = 0; headX = 0; headY = 0;
  stimChannel = 0; stimParam = 0; stimStrength = 0; stimUntil = 0; totalSpikes = 0;
  onStep: ((spikes: number[], sn: number) => void) | null = null;
  constructor(circuit: Circuit, params: Params) {
    this.c = circuit; this.p = params;
    const N = circuit.N;
    this.v = new Array(N).fill(0); this.bias = new Array(N).fill(0); this.inp = new Array(N).fill(0); this.hist = new Array(WEDGES).fill(0);
  }
  loadState(s: ChainState) {
    this.v = s.v.map(Number); this.bias = s.bias.map(Number); this.hist = s.hist.map(Number);
    this.inp = s.pendingInput ? s.pendingInput.map(Number) : new Array(this.c.N).fill(0);
    this.step = Number(s.step); this.energy = Number(s.energy); this.alive = !!s.alive; this.generation = Number(s.generation);
    this.posX = Number(s.posX); this.posY = Number(s.posY); this.headX = Number(s.headX); this.headY = Number(s.headY);
    if (s.stim) { this.stimChannel = Number(s.stim.channel); this.stimParam = Number(s.stim.param); this.stimStrength = Number(s.stim.strength); this.stimUntil = Number(s.stim.untilStep); }
  }
  stimulate(channel: number, param: number, strength: number) { this.stimChannel = channel; this.stimParam = param; this.stimStrength = strength; this.stimUntil = this.step + this.p.stimTTL; }
  buildStim() {
    const c = this.c, p = this.p, amp = this.stimStrength * p.stimGain, stim = new Array(c.N).fill(0);
    const ch = this.stimChannel, param = this.stimParam;
    for (let i = 0; i < c.N; i++) {
      const t = c.type[i];
      if (ch === CH.CUE) { if (t <= T.EPGT && c.wedge[i] !== 255) { const dist = (c.wedge[i] + WEDGES - param) % WEDGES; if (dist === 0) stim[i] = amp; else if (dist === 1 || dist === WEDGES - 1) stim[i] = tdiv(amp, 2); } }
      else if (ch === CH.TURN_LEFT || ch === CH.TURN_RIGHT) { if ((t === T.PEN_A || t === T.PEN_B) && ((ch === CH.TURN_LEFT && c.side[i] === 0) || (ch === CH.TURN_RIGHT && c.side[i] === 1))) stim[i] = amp; }
      else if (ch === CH.SHOCK) { if (t === T.DELTA7) stim[i] = amp; }
    }
    return stim;
  }
  tick(steps: number) {
    const c = this.c, p = this.p, N = c.N;
    if (!this.alive) return null;
    const stimActive = this.stimChannel !== 0 && this.step < this.stimUntil;
    const stimI = stimActive ? this.buildStim() : null;
    if (!p.persistInput) this.inp.fill(0);
    const spk = new Array(N).fill(0), bins = new Array(WEDGES).fill(0);
    let hx = 0, hy = 0, tickSpikes = 0, ran = 0;
    const s0 = this.step;
    while (ran < steps && this.energy > 0) {
      const sn = s0 + ran, stimNow = stimActive && sn < this.stimUntil;
      let rnd = keccakU64(sn);
      const spikeList: number[] = [];
      for (let i = 0; i < N; i++) {
        if ((i & 31) === 0 && i !== 0) rnd = keccakU256(rnd);
        const nz = Number((rnd >> BigInt((i & 31) * 8)) & 0xffn) - 128;
        let x = this.v[i];
        x -= tdiv(x * p.leak, 1024);
        x += this.inp[i] + tdiv(nz * p.noise, 128) + this.bias[i] * p.gBias;
        if (stimNow) x += stimI![i];
        this.inp[i] = 0;
        if (x < p.vMin) x = p.vMin;
        if (x >= p.thresh) { x = p.reset; spikeList.push(i); spk[i]++; if (c.type[i] <= T.EPGT && c.wedge[i] !== 255) { hx += COS16[c.wedge[i]]; hy += SIN16[c.wedge[i]]; bins[c.wedge[i]]++; } }
        this.v[i] = x;
      }
      tickSpikes += spikeList.length;
      for (const i of spikeList) { const g = p.gains[c.type[i]]; for (const [post, w] of c.out[i]) this.inp[post] += tdiv(w * g, 16); }
      if (this.onStep) this.onStep(spikeList, sn);
      ran++; this.energy--;
    }
    for (let i = 0; i < N; i++) { let b = this.bias[i]; if (spk[i] * 8 >= ran) { if (b < BIAS_MAX) b++; } else if (spk[i] === 0) { if (b > -BIAS_MAX) b--; } this.bias[i] = b; }
    let best = 0, bestCount = 0; for (let w = 0; w < WEDGES; w++) if (bins[w] > bestCount) { bestCount = bins[w]; best = w; }
    if (bestCount > 0 && this.hist[best] < 0xffff) this.hist[best]++;
    const mag = isqrt(hx * hx + hy * hy);
    if (mag > 0 && mag >= p.walkThreshold * ran) { this.posX += tdiv(hx * STRIDE * ran, mag); this.posY += tdiv(hy * STRIDE * ran, mag); }
    this.headX = hx; this.headY = hy; this.step = s0 + ran; this.totalSpikes += tickSpikes;
    if (stimActive && this.step >= this.stimUntil) this.stimChannel = 0;
    if (this.energy === 0) this.alive = false;
    return { steps: ran, spikes: tickSpikes, spk, bins, hx, hy };
  }
}
